### PR TITLE
feat: ims client (WIP)

### PR DESCRIPTION
**Pull request by solaris007**
IMS Client

Blocks https://github.com/adobe/spacecat-fulfillment-worker/pull/2

### CODE REVIEW

File: packages/spacecat-shared-ims-client/src/clients/ims-client.js [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+      }
+    }
+
+    // Transform the object "map" back to a de-duped array
+    return Object.keys(users).map((email) => users[email]);
+  }
+
+  async getServiceAccessToken() {
+    if (hasText(this.serviceAccessToken)) {
+      return this.serviceAccessToken;
```
Comment (ekremney) [inline packages/spacecat-shared-ims-client/src/clients/ims-client.js:185]: heads-up: this function returns a string here but down below it returns an object

File: packages/spacecat-shared-ims-client/src/clients/ims-client.js [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+      );
+
+      this.#logDuration('IMS getServiceAccessToken request', startTime);
+
+      if (!tokenResponse.ok) {
+        throw new Error(`IMS getServiceAccessToken request failed with status: ${tokenResponse.status}`);
+      }
+
+      /* eslint-disable camelcase */
+      const { access_token, token_type, expires_in } = await tokenResponse.json();
```
Comment (ekremney) [inline packages/spacecat-shared-ims-client/src/clients/ims-client.js:211]: feels like this is missing here:

```
this.serviceAccessToken = access_token;
```

File: packages/spacecat-shared-ims-client/src/clients/ims-client.js [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+      this.serviceAccessToken = {
+        access_token,
+        expires_in,
+        token_type,
+      };
+
+      return this.serviceAccessToken;
+    } catch (error) {
```
Comment (blefebvre) [inline packages/spacecat-shared-ims-client/src/clients/ims-client.js:220]: Is this catch block necessary? An error fetching a token would be thrown above (line 207) then re-thrown in the catch.

File: packages/spacecat-shared-ims-client/src/clients/ims-client.js [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    const productContext = await this.#getProductContextByImsOrgId(imsOrgId);
+    const tenantId = productContext?.params?.tenant_id;
+
+    // Request organization details
+    const orgDetails = await this.#getImsOrgDetails(imsOrgId);
+    const orgName = orgDetails?.orgName;
+
+    this.log.debug(`IMS Org ID ${imsOrgId} has tenantId: ${tenantId}, name: "${orgName}"`);
+
+    // Request users from all the known IMS groups for this Org
```
Comment (blefebvre) [inline packages/spacecat-shared-ims-client/src/clients/ims-client.js:241]:
```suggestion
    // Fetch a list of all users in the administrators group
```

File: packages/spacecat-shared-ims-client/src/clients/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+export class ImsClient {
+  /**
+   * Creates a new ImsClient instance from the given UniversalContext.
+   * @param {UniversalContext} context The UniversalContext to use for creating the ImsClient.
+   * @returns {ImsClient} The ImsClient instance.
+   */
+  static createFrom(context: UniversalContext): ImsClient;
+
+  /**
+   * Returns the access token for the given scope.
```
Comment (blefebvre) [inline packages/spacecat-shared-ims-client/src/clients/index.d.ts:24]:
```suggestion
   * Returns an access token for the scopes associated with the IMS client ID.
```

File: packages/spacecat-shared-ims-client/src/utils.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+ * Generates the IMS organizations endpoint URL.
+ * @param {string} imsHost - The IMS host.
+ * @return {`/ims/organizations/${string}/v2`} - The IMS organizations endpoint URL.
+ */
+export const getImsOrgsApiPath = (imsOrgId) => `${IMS_ORGANIZATIONS_ENDPOINT}/${imsOrgId}/v2`;
+
+/**
+ * Extracts the orgId and authSource from the IMS Org ID.
+ * @param {string} imsOrgId - The IMS Org ID.
+ * @return {{authSource: *, orgId: *}} - The orgId and authSource.
```
Comment (blefebvre) [inline packages/spacecat-shared-ims-client/src/utils.js:54]:
```suggestion
 * @return {{authSource: string, orgId: string}} - The orgId and authSource.
```

**Review by blefebvre — APPROVED** [review]
👍 with some minor suggestions.

File: packages/spacecat-shared-ims-client/src/clients/ims-client.js [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        }
+      }
+    }
+
+    // Transform the object "map" back to a de-duped array
+    return Object.keys(users).map((email) => users[email]);
+  }
+
+  async getServiceAccessToken() {
+    if (hasText(this.serviceAccessToken?.access_token)) {
```
Comment (ekremney) [inline packages/spacecat-shared-ims-client/src/clients/ims-client.js:184]: what if the token `hasText` but expired? 👀

File: packages/spacecat-shared-ims-client/src/clients/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import type { UniversalContext } from '@adobe/helix-universal';
+
+export class ImsClient {
```
Comment (ekremney) [inline packages/spacecat-shared-ims-client/src/clients/index.d.ts:15]: 💯

**Review by ekremney — APPROVED** [review]

**Pull request merged**